### ISSUE TITLE
Avoid deadlocks in the crawler by doing address book updates and handshakes in separate tasks

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#3288]
## Motivation

Zebra can deadlock when `CandidateSet::update` is waiting for ready peers to get more peers, but the PeerSet is waiting for the crawler to send new peers.

### Scheduling

We should only do this task if:
- we've fixed all the known hang bugs, and
- Zebra still has frequent hangs.

This code has had multiple hang bugs before, so we should schedule this fix first, if we need hang fixes.

### Designs

Choose one or more of these alternatives:
- pass the `PeerSet` update channel into each handshake task, so new peers don't block on the rest of the crawler tasks
- spawn each `CandidateSet::update` request as a separate task, and wait on their `JoinHandle`s in the main crawler loop, so address book updates don't block on the rest of the crawler tasks
- Test for hangs with no available peer set or address book peers

### Testing

Zebra can hang when:
- there are no ready `PeerSet` peers,
- the `PeerSet` is empty,
- there are no `AddressBook` peers available for reconnection,
- all `AddressBook` peers will never be used for reconnection, or
- the `AddressBook` is empty.

Or some combination of these situations.

We should add tests for these hangs, so we don't accidentally make them worse.

**Comment by ftm1000** [comment]
redistributing issues that can be separately worked from Epic #2311

**Comment by ftm1000** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Engineering-Team-5fb24d9264a3e8000e666a9e/issues/ZcashFoundation/zebra/3288?planning-poker) with ZenHub @conradoplg @dconnolly @gustavovalverde @jvff @oxarbitrage @teor2345 @upbqdn

**Comment by teor2345** [comment]
This isn't actually causing any deadlocks.

**Comment by teor2345** [comment]
We didn't actually fix this, it might be causing issues now

**Comment by upbqdn** [comment]
@teor2345, you closed this right after reopening it, so I wonder if it was by accident.

**Comment by teor2345** [comment]
> @teor2345, you closed this right after reopening it, so I wonder if it was by accident.

Thanks for checking!

It was marked as "completed", which was confusing because we never fixed it. I handed it over to Arya yesterday, so it's up to him to decide if this issue is the cause of the hangs we're seeing.

**Comment by teor2345** [comment]
We ended up fixing this in PR #6869
